runners/prism/java/src/main/java/org/apache/beam/runners/prism/PrismExecutor.java

```diff
@@ -0,0 +1,199 @@
+
+  protected @MonotonicNonNull Process process;
+  protected ExecutorService executorService = Executors.newSingleThreadExecutor();
+  protected @MonotonicNonNull Future<?> future = null;
+
+  static Builder builder() {
+    return new AutoValue_PrismExecutor.Builder();
+  }
+
+  /** The command to execute the Prism binary. */
+  abstract String getCommand();
+
+  /**
+   * Additional arguments to pass when invoking the Prism binary. Defaults to an {@link
+   * Collections#emptyList()}.
+   */
+  abstract List<String> getArguments();
+
+  /** Stops the execution of the {@link Process}, created as a result of {@link #execute}. */
+  void stop() {
+    LOG.info("Stopping Prism...");
+    if (future != null) {
+      future.cancel(true);
+    }
+    executorService.shutdown();
+    try {
+      boolean ignored = executorService.awaitTermination(1000L, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ignored) {
+    }
+    if (process == null) {
+      return;
+    }
+    if (!process.isAlive()) {
+      return;
+    }
+    process.destroy();
+    try {
+      process.waitFor();
+    } catch (InterruptedException ignored) {
+    }
+  }
+
+  /**
+   * Execute the {@link ProcessBuilder} that starts the Prism service. Redirects output to STDOUT.
+   */
+  void execute() throws IOException {
+    execute(createProcessBuilder().inheritIO());
+  }
+
+  /**
+   * Execute the {@link ProcessBuilder} that starts the Prism service. Redirects output to the
+   * {@param outputStream}.
+   */
+  void execute(OutputStream outputStream) throws IOException {
+    execute(createProcessBuilder().redirectErrorStream(true));
+    this.future =
+        executorService.submit(
+            () -> {
+              try {
+                ByteStreams.copy(checkStateNotNull(process).getInputStream(), outputStream);
+              } catch (IOException e) {
+                throw new RuntimeException(e);
+              }
+            });
+  }
+
+  /**
+   * Execute the {@link ProcessBuilder} that starts the Prism service. Redirects output to the
+   * {@param file}.
+   */
+  void execute(File file) throws IOException {
+    execute(
+        createProcessBuilder()
+            .redirectErrorStream(true)
+            .redirectOutput(ProcessBuilder.Redirect.appendTo(file)));
+  }
+
+  private void execute(ProcessBuilder processBuilder) throws IOException {
+    this.process = processBuilder.start();
+    LOG.info(
+        "started {}, pid: {}",
+        String.join(" ", getCommandWithArguments()),
+        getPidOrNotAvailableWarning());
+  }
+
+  private List<String> getCommandWithArguments() {
+    List<String> commandWithArguments = new ArrayList<>();
+    commandWithArguments.add(getCommand());
+    commandWithArguments.addAll(getArguments());
+
+    return commandWithArguments;
+  }
+
+  private ProcessBuilder createProcessBuilder() {
+    return new ProcessBuilder(getCommandWithArguments());
+  }
+
+  private String getPidOrNotAvailableWarning() {
```

Comment:
these three private methods look very strange to me. Why need reflection to get a pid from a process? Doesn't process.pid() work?

Reply:
@Abacn Not all Java versions have support for querying the PID. The recommendation is to use reflection to determine whether support exists.

Reply:
I see, https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/lang/Process.html#pid-- says "Since:
9".

Would you mind adding a comment and also a TODO saying moving on to pid() when beam/java Prism runner drops Java8 support?